src/csharp/Grpc.IntegrationTesting/TestGrpc.cs

```diff
@@ -34,7 +34,15 @@ public static partial class TestService
   {
     static readonly string __ServiceName = "grpc.testing.TestService";
 
-    static readonly grpc::Marshaller<global::Grpc.Testing.Empty> __Marshaller_grpc_testing_Empty = grpc::Marshallers.Create((arg) => global::Google.Protobuf.MessageExtensions.ToByteArray(arg), global::Grpc.Testing.Empty.Parser.ParseFrom);
+    static readonly grpc::Marshaller<global::Grpc.Testing.Empty> __Marshaller_grpc_testing_Empty = grpc::Marshallers.Create(
+      (arg) => global::Google.Protobuf.MessageExtensions.ToByteArray(arg),
+      global::Grpc.Testing.Empty.Parser.ParseFrom
+      (output, bufferWriter) =>
+      {
+          var writer = new CodedOutputWriter(bufferWriter);
+          output.WriteTo(ref writer);
+      },
+      global::Grpc.Testing.Empty.Parser.ParseFrom);
```

Comment:
IMHO the simplest possible approach is to generate code like this:
```
grpc::Marshallers.Create( (msg, ctx) =>
{
  // serialization
  ctx.SetPayloadLength(msg.CalculateSize());
  msg.writeTo(ctx.GetBufferWriter());
  ctx.Complete();
},
(ctx) =>
{
  return global::Math.DivArgs.Parser.ParseFrom(ctx.PayloadAsReadOnlySequence());
});
```

That doesn't require the "BufferMarshaller" class at all.

At the same time, I was wondering if we can come up with a more "intelligent" solution that would choose the right serialization approach depending on whether the message type supports IBufferMessage etc.
(so that we opt for faster code path automatically if it's available).

Reply:
We can make the generated code do a runtime check for `IBufferMessage` and then choose between this and converting to a byte array. That is what Google.Protobuf already does for non-root messages.

I'm not sure how much it would be needed here though. Messages will typically be generated at the same time as the service, so we can rely on them being buffer serializable messages. Also, known types that aren't generated at build should already support buffer serialization in latest versions of Google.Protobuf.

Selectively turning this on for one or all services is always an option for devs if there are methods that have problematic messages..

Reply:
I considered doing a runtime check but I was wondering if the generated code is the best place to do it. I think we should add the check here because there's a corner case where messages are built separately and do not not support buffer serialization. I didn't want to regress this scenario, though I don't think it's a common one. I'll update the generated code.

Reply:
> That is what Google.Protobuf already does for non-root messages.

Not sure I follow, can you point me to the code? I thought that whenever we start using CodedInputReader parsing for the root message, we use it all the way for all the submessages also (and we expect the generated methods to exist). Same for serialization.